### PR TITLE
Add exhaustivity check tests

### PR DESCRIPTION
**Pull request by liufengyun**
Add exhaustivity check tests

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -1020,6 +1020,12 @@ object Types {
       case _ => this
     }
 
+    /** Widen singleton type modulo constant types */
+    final def widenNonConstant(implicit ctx: Context): Type = this match {
+      case _: ConstantType => this
+      case _ => widen
+    }
+
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: It seems to me better to revert this change, as  (1) `tests/patmat/t11603.scala` is not a real use case and it's hard to imagine one; (2) it does not work if the scrutinee and pattern are both singletons (again why use patmat for such case).

File: tests/patmat/t11620.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  case B.Aux(a @ A2(_, _)) ⇒ a.t1  // 👎 (false-positive): unreachable code
+}
+
+def foo3[_A[+U] <: A[U], T](b: B.Aux[_A, T]) = b match {
+  case B.Aux(a: A1[T]) ⇒ a.t
+  case B.Aux(a: A2[T]) ⇒ a.t1  // 👎 (false-positive): unreachable code
+}
+
+def foo4[T](b: B[T]) = b match {
+  case B(A1(t)) ⇒ t  // 👎 (false-negative): incomplete match
```
Comment (nicolasstucki) [inline tests/patmat/t11620.scala]: Do we have issues for these false-negative an false-positive cases?

Comment (liufengyun) [reply tests/patmat/t11620.scala]: Here, a warning is expected, and the check file `tests/patmat/t11620.check` checks that at line 36 there should be a warning:

```
36: Pattern Match Exhaustivity: B(A2(_, _))
```
`
